### PR TITLE
Add boolean flag to skip argument type checking of builtin functions.

### LINKED ISSUE #228: make log work with any types
currently the `log` function signature is hardcoded as:

```rust
const LOG_FN: &str = "log(var: Field)";
```

this makes the type checker forbid the function from being used with other types than `Field`.

To fix that, we need to:

1. make it so that the arguments part of a hardcoded builtin signature can be ignored by the type checker
2. while still making sure that the type checker can verify that the returned type of a builtin function matches the expected return type expected by the caller (the builtin function does not have that information so cannot do that type checking)

In `get_parsed_fns` we need a flag that signals this:

```rust
        for (code, fn_handle, ignore_args) in fns {
```

Then `FnKind::BuiltIn` needs to carry that information as well.

Finally, in the typechecker, we can ignore type checks on the function arguments if we see this flag.

### PR DESCRIPTION
**Pull request by bufferhe4d**
Attempts to resolve #228.

Here is my approach:

As suggested in #228 I first introduced a new flag for the `FnKind::Builtin` called `ignore_arg_types`:

https://github.com/bufferhe4d/noname/blob/b243dce51ddbb96babf627fd3c1a8b5266d261ab/src/imports.rs#L78

Propagated this to all the occurrences of this. Then implemented the actual skip in the type checker.

### CODE REVIEW

File: src/type_checker/checker.rs
Comment (bufferhe4d) [inline src/type_checker/checker.rs:839]: Seems to work for me but I am not sure if this is the correct way to check if a function is builtin.

File: src/type_checker/mod.rs
Comment (bufferhe4d) [inline src/type_checker/mod.rs:396]: I couldn't immediately decide what to do here since we first convert a hint function to a builtin one. It felt like we would never disable argument type checking for a hint function so I defaulted to false.

Comment (katat) [reply src/type_checker/mod.rs:396]: The builtin hints will be deprecated soon. So don't worry bout this. 
Good insight!

**Review by katat — APPROVED** [review]
lgtm!
@mimoo may want another pass

**Comment by mimoo** [comment]
maybe we should check if this allows us to implement what we wanted to implement on top of that?

@varunthakore do you want to try to implement a generic `log()` function on top of this branch to see if it works?

@bufferhe4d if you want to do something similar, a generic `assert_eq()` function would be great, as long as it checks that the two types passed as arguments match

**Comment by bufferhe4d** [comment]
> @bufferhe4d if you want to do something similar, a generic `assert_eq()` function would be great, as long as it checks that the two types passed as arguments match

I currently have an assert_eq that is working with all the types (Field, Bool, a7rbitrary depth structs and arrays) in the language (using the boolean flag we introduced in this PR). It only checks whether the two argument types matched as you described.

Since it requires the flag included here, should I push this functionality to this branch to make it a part of this PR or should I wait until this is merged to open a separate PR?

**Comment by mimoo** [comment]
can you open a separate PR that targets this branch? Or another PR that we can easily rebase?

**Comment by bufferhe4d** [comment]
I created a draft PR from my generic_assert_eq here: https://github.com/bufferhe4d/noname/pull/2

The CI should be fixed now.

**Pull request merged**
